app/components/RuleViewerEditor/subcomponents/RuleInputOutputFieldsComponent.tsx

```diff
@@ -0,0 +1,197 @@
+import { SchemaSelectProps } from "@gorules/jdm-editor/src/helpers/components";
+import { KlammBREField } from "@/app/types/klamm";
+import { getBREFields } from "@/app/utils/api";
+import styles from "./RuleInputOutputFieldsComponent.module.css";
+
+declare type InputOutputField = {
+  id?: string;
+  field: string;
+  label?: string;
+  type?: string;
+};
+
+interface RuleInputOutputFieldsComponent extends GraphNodeProps {
+  setInputOutputSchema: (schema: SchemaSelectProps[]) => void;
+  isEditable: boolean;
+}
+
+export default function RuleInputOutputFieldsComponent({
+  specification,
+  id,
+  isSelected,
+  name,
+  setInputOutputSchema,
+  isEditable,
+}: RuleInputOutputFieldsComponent) {
+  const { updateNode } = useDecisionGraphActions();
+  const node = useDecisionGraphState((state) => (state.decisionGraph?.nodes || []).find((n) => n.id === id));
+  const inputOutputFields: InputOutputField[] = node?.content?.fields || [];
+
+  const [inputOutputOptions, setInputOutputOptions] = useState<BaseOptionType[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    const getFieldsFromKlamm = async () => {
+      try {
+        const data: KlammBREField[] = await getBREFields();
+        const newInputOutputOptions: BaseOptionType[] = data.map(({ id, name, label, description }) => ({
+          label: `${label}${description ? `: ${description}` : ""}`, // Add the description as part of the label - will be formatted properly later
+          value: name,
+        }));
+        setInputOutputOptions(newInputOutputOptions);
+        setIsLoading(false);
+      } catch (error) {
+        console.error("Error:", error);
+      }
+    };
+    getFieldsFromKlamm();
+  }, []);
+
+  useEffect(() => {
+    // Add a new field by default if one doesn't exist when editing
+    if (isEditable && inputOutputFields?.length == 0) {
+      updateNode(id, (draft) => {
+        draft.content = { fields: [{ id: crypto.randomUUID(), field: "" }] };
+        return draft;
+      });
+    }
+  }, []);
+
+  useEffect(() => {
+    if (inputOutputFields?.length == 0) return;
+    // Map the fields from the schema - basically just converting label to name
+    const schemafiedInputs = inputOutputFields
+      .filter(({ field }) => field)
+      .map(({ field, label }) => ({ field, name: label }));
+    setInputOutputSchema(schemafiedInputs);
+  }, [inputOutputFields]);
+
+  const addInputField = () => {
+    updateNode(id, (draft) => {
+      if (!draft.content?.fields) {
+        draft.content = { fields: [] };
+      }
+      draft.content.fields.push({ id: crypto.randomUUID(), field: "" });
+      return draft;
+    });
+  };
+
+  const updateInputField = (item: InputOutputField, { key, label }: any) => {
+    updateNode(id, (draft) => {
+      draft.content.fields = draft.content.fields.map((input: InputOutputField) => {
+        if (input.id === item.id) {
+          input.field = key;
+          input.label = label;
```

Comment:
I think you already mentioned pulling from the api for the input fields, so would we want to add the KLAMM id for the field inputs saved directly into the rule schema? Ideas on what the best approach would be for keeping this in sync?

Reply:
Yeah, need to figure out what the best thing to store actually is. Since it's being stored on the file itself, might want to keep it rather small. Could do just the field and the id. That would pretty much keep everyone in sync the easiest, but it would be a lot of fetching. Might be best to store the field, id, name, data_type, and data_validation so we have information we could use whenever a rule is run without having to fetch additional info every time. But then we run into potential syncing problems you've alluded too. 

I think we'd have to come up with a way of automatically re-syncing this information when needed. This could be done during the rule creation/editing process potentially. So like the Klamm information for a rule wouldn't update until a user goes in and edits it in some way. This would also have the potential benefit of not needing to rely on Klamm at all once a rule is published. But it could potentially be annoying if the type or validation of an input that is used across many rules is updated in Klamm. Of course we could create additional bulk processes that do this if it becomes a common thing.

Thoughts?

Reply:
I agree that storing the basic field, id, name etc would be the most efficient and space saving.

I really liked your previous implementations of github pr's with the bre frontend edits, and I wonder if it would be possible to implement something similar with KLAMM? I'm unsure on the technical difficulty, especially with laravel, but could we explore the option of when a field is changed in some way in KLAMM, it propagates as a change to all relevant files in the repository? 

https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content

On a KLAMM edit, the rules repo could be queried and updated, returning all files that reference the respective field based on the relevant id, and updating the files with this data?

Reply:
Yeah, I think that should be possible but it does seem like a bit of a headache. I think we should hold that idea as a possible future enhancement. It might become a necessity but I kinda want to see first.